### PR TITLE
kernel/hil/time,capsules/alarm: pad Ticks to 32 bit for predictable wrapping

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This change introduces new APIs on the `Ticks` trait of `kernel/src/hil/time.rs` which allow consumers of this type to know the underlying timer's actual ticks width (may be less than 32 bit), and to right-pad Ticks values for it to wrap at exactly `(2 ** 32) - 1` ticks.

By exposing this shifted Ticks value and an appropriately scaled tick frequency to userspace, Tock finally conforms to assumptions made in `libtock-c` long ago:

     The client should not assume anything about the underlying clock
     used by an implementation other than that it is running at
     sufficient frequency to deliver at least millisecond granularity
     and that it is a 32-bit clock (i.e. it will wrap at 2^32 clock
     ticks).

This predictable timer interface can allow userspace to build timer infrastructure that can schedule alarms or sleep for longer than `2 ** Timer Width` ticks.

Apart from now conforming to libtock-c's expectations, by scaling both timer and frequency this change should not break any existing userspace applications (apart from triggering an otherwise unrelated bug in libtock-c's virtual alarms). It does modify code for which a relevant TRD (105) exists, but that TRD is still in draft state.

### Testing Strategy

Needs more testing. Will test along with some changes to `libtock-c` tomorrow.

Needs draft TRD update.

### TODO or Help Wanted

N/A

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
Probably should add the usize_padded function too.

**Comment by bradjc** [comment]
And `_left_justified` might be a more specific name.

File: kernel/src/hil/time.rs [hunk 1]
```diff
@@ -35,6 +42,33 @@ pub trait Ticks: Clone + Copy + From<u32> + fmt::Debug + Ord + PartialOrd + Eq {
     /// are 32 bits.
     fn into_u32(self) -> u32;
 
+    /// The amount of bits required to right-pad this timer (filling
+    /// the lower bits with `0`) for it wrap at `(2 ** 32) - 1`
+    /// bits. For timers with a `width` larger than 32, this value
+    /// will be `0` (i.e., they can simply be truncated to 32-bits).
+    fn u32_padding() -> usize {
+        32_usize.saturating_sub(Self::width())
```
Comment (brghena) [inline kernel/src/hil/time.rs:50]: (pedantic comment) I think this should probably be `u32::BITS` https://doc.rust-lang.org/core/primitive.u32.html#associatedconstant.BITS

**Review by Samir-Rashid — COMMENTED** [review]
I do not understand the fix that is being done in this code or why it fixes the problem. I think the following approach, which was mentioned on a core team call, is better.

### I think the problem (root cause of this change) is

The NRF 52840dk has a 24 bit counter. At a frequency of 32.768kHz, this buffer overflows every  $(2^{24} -1) / 32768 = 512$ seconds. This bug was discovered from the OpenThread porting, any alarm you set from `alarm_at` or `timer_in` is always going to incorrectly fire within 512 seconds.

### I thought the solution is

You take the 24 bit prescaler and glue it to the 12 bits you have in the prescaler. So this becomes `| 24 bits COUNTER | 8 bits from PRESCALER |`. You want the counter to be the MSBs because that triggers the overflow interrupt. Then to get the rest of the bits, you use the prescaler as the LSBs. This is the same as having a continuous 32 bit buffer because the prescaler will increment the counter buffer when it overflows in hardware.

image

Thus our counter becomes actually 32 bits.

Pseudocode:
```c
set_prescaler_to_8_bits(); // prescaler should fire its overflow at 8 bits
uint24_t counter_value = get_counter();
uint8_t prescaler_value = get_prescaler();
return (uint32_t) ((counter_value << 8) & prescaler_value);
```

### This PR

I understand the need for this change is to fix that the kernel currently is lying about having the timer "wrap at 2^32 clock ticks". Purely in this sense, this PR ought to get merged. However, I foresee there being big issues with such a large frequency. For the NRF52840dk, the new "frequency" that the kernel is going to tell userspace is $32768 << 8 = 2^{23} = 8388608$. This is a huge number, and I believe it is unnatural for a developer to have to deal with this.

Furthermore, this padding does not affect the problem that this counter buffer is still going to overflow less than every 10 minutes. Why was the above method of combining prescaler and counter buffers not chosen? I believe that approach better addresses the original overflow concern, also brings the kernel into compatibility with the 32 bit overflow claim, is a concise fix, is better for userspace applications, and is more power efficient due to less overflows.

**Comment by bradjc** [comment]
8 MHz seems relatively small all things considered. A SPI bus can easily run at a higher frequency, and we definitely want our CPU clocks to run faster. I don't see why developers would have any issue with that frequency.

But in any case, these seem like orthogonal changes. It is very possible that a platform/use case needs the resolution of the clock without the prescaler, or that a different board only has a 24 bit counter without a prescaler. In those cases, having a fixed 32 bit counter is still valuable.

**Comment by Samir-Rashid** [comment]
If it is not standard for boards to have 32 bits among the counter+prescaler, then I agree.

Multiplying 8Mhz by >511 will overflow a 32 bit integer. This PR accelerates the need for all alarm stuff to happen in 64 bits.

**Comment by Samir-Rashid** [comment]
My mistake, fixed. The upper bound on error is 1 **milli**second, not **micro**second.

**Review by lschuermann — COMMENTED** [review]
I integrated your requested changes @brghena and @bradjc. Let me know what you think.

Concerning this larger discussion, I see @Samir-Rashid's argument, but I do think I agree with @bradjc here. I strongly believe it is important for Tock to be able to support timers < 24 bit. Those are relatively prevalent across hardware, especially in low-energy systems that want to avoid large timers. Thus this logic is necessary for us to provide a reasonable system call interface with strong semantic guarantees, as is evident by `libtock-c`'s reliance on exactly these guarantees.

However, I believe there are still two remaining issues with this changeset:

- I do agree that this approach does not work for all timers, widths and frequencies. In particular, once we hit a scaled / left-justified timer frequency `> u32::MAX`, this code will break. For those systems, the kernel would need to emulate a larger timer by means of implementing a wraparound.

  This changeset is sub-optimal, as it does not clearly express these constraints, neither in the Time HIL, nor in the capsule instantiation.

- I am having a hard time convincing myself that the wrapping logic in the alarm capsule correctly handles timers with widths of less than 32 bit. In particular, `reset_active_alarm` seems problematic, as it assumes the underlying timer to be at least 32 bit wide.
  https://github.com/tock/tock/blob/984bc400d1b618e5e8390038f5a7ba42ce604c09/capsules/core/src/alarm.rs#L63-L155
  
  These problems are exacerbated by a lack of test-ability outside of a full kernel, as the timer logic is intertwined with setting timers and issuing callbacks.

Currently, the alarm architecture performs conversions along the lines of `32 bit (scaled) -> 24 bit -> 32 bit (unscaled) -> A::Ticks::width bit`. Taking the lessons learned over the past few weeks into account, it seems that this should instead be `32 bit (scaled) -> A::Ticks::width bit`, and thus:
- make the alarm wrapping logic fully generic over the underlying timer width, by using the appropriate functions of the Time HIL,
- up- and down-converting converting `dt` and `reference` to the appropriate timer width only in the system call interface.

These changes are more invasive, I'll see whether I can get something workable up today.

**Review by bradjc — APPROVED** [review]

**Comment by lschuermann** [comment]
#3975 implements the changes I talked about above. I'm pretty happy with the outcome. I acknowledge that a full rewrite is not as easy to review as this changeset, but I think that the approach taken in #3975 is better than this hot-patch here. That being said, I'm happy to have this merged in the meantime, while we discuss and review #3975.

**Comment by tyler-potyondy** [comment]
Is this something we want to merge now? From the libtock/openthread perspective, this PR provides the needed left alignment to resolve the timer wrapping bug (as Leon pointed out, potentially at the cost of introducing new bugs to the alarm capsule given the difficulty reasoning about this and challenges testing the subsystem prior to Leon's changes in #3975).

**Comment by lschuermann** [comment]
I think it'd be really cool if we can merge #3975 instead, if reviewers are confident enough in these changes. If not let's go with this one.

**Pull request merged**